src/run.rs

```diff
@@ -0,0 +1,188 @@
+    d.file_type().is_dir()
+}
+
+/// Run all functions in a file that are succeeded by "run:" comments
+fn run_single_file(path: &PathBuf) -> Result<(), String> {
+    let file_contents = read_to_string(&path).map_err(|e| e.to_string())?;
+    run_file_contents(file_contents)
+}
+
+/// Main body of `run_single_file` separated for testing
+fn run_file_contents(file_contents: String) -> Result<(), String> {
+    let test_file = parse_test(&file_contents, None, None).map_err(|e| e.to_string())?;
+    let isa = create_target_isa(test_file.isa_spec)?;
+    for (func, Details { comments, .. }) in test_file.functions {
+        if comments.iter().filter(|c| c.text.contains("run")).count() > 0 {
+            run_single_function(func, isa.as_ref())?
+        }
+    }
+    Ok(())
+}
+
+/// Build an ISA based on the current machine running this code (the host)
+fn create_target_isa(isa_spec: IsaSpec) -> Result<Box<dyn TargetIsa>, String> {
+    if let IsaSpec::None(flags) = isa_spec {
+        // build an ISA for the current machine
+        let builder = host_isa_builder()?;
+        Ok(builder.finish(flags))
+    } else {
+        Err(String::from("A target ISA was specified in the file but should not have been--only the host ISA can be used for running CLIF files"))?
+    }
+}
+
+/// Compile and execute a single function, expecting a boolean to be returned; a 'true' value is
+/// interpreted as a successful test execution and mapped to Ok whereas a 'false' value is
+/// interpreted as a failed test and mapped to Err.
+fn run_single_function(func: Function, isa: &dyn TargetIsa) -> Result<(), String> {
+    if !(func.signature.params.is_empty()
+        && func.signature.returns.len() == 1
+        && func.signature.returns.first().unwrap().value_type.is_bool())
+    {
+        return Err(String::from(
+            "Functions must have a signature like: () -> boolean",
+        ));
+    }
+    // TODO observed segfaults with return types of b64
+
+    // set up the context
+    let mut context = Context::new();
+    context.func = func;
+
+    // compile and encode the result to machine code
+    let relocs = &mut NullRelocSink {};
+    let traps = &mut NullTrapSink {};
+    let stackmaps = &mut NullStackmapSink {};
+    let code_info = context.compile(isa).map_err(|e| e.to_string())?;
+    let code_page = MemoryMap::new(code_info.total_size as usize, &[MapOption::MapWritable])
+        .map_err(|e| e.to_string())?;
+    let callable_fn: fn() -> bool = unsafe {
+        context.emit_to_memory(isa, code_page.data(), relocs, traps, stackmaps);
+        region::protect(code_page.data(), code_page.len(), Protection::ReadExecute)
+            .map_err(|e| e.to_string())?;
+        mem::transmute(code_page.data())
+    };
+
+    // execute
+    match callable_fn() {
+        true => Ok(()),
+        false => Err(format!("Failed: {}", context.func.name.to_string())),
+    }
+}
+
+/// Relocation helper
+struct NullRelocSink {}
```

Comment:
All these null sinks already exist in codegen::binemit (or elsewhere for the other sinks), could they be imported instead?